sdk/core/core/src/az_json_read.c

```diff
@@ -0,0 +1,376 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include <az_json_read.h>
+
+#include <az_digit.h>
+
+#include <math.h>
+
+inline bool az_json_is_white_space(char const c) {
+  switch (c) {
+    case ' ': case '\t': case '\n': case '\r':
+      return true;
+  }
+  return false;
+}
+
+inline bool az_json_is_esc(char const c) {
+  switch (c) {
+    case '\\':
+    case '"':
+    case '/':
+    case 'b':
+    case 'f':
+    case 'n':
+    case 'r':
+    case 't':
+      return true;
+  }
+  return false;
+}
+
+inline bool az_json_stack_is_empty(az_json_state const *const p) {
+  return p->stack == 1;
+}
+
+inline az_json_stack_item az_json_stack_last(az_json_state const *const p) {
+  return p->stack & 1;
+}
+
+az_json_state az_json_state_create(az_const_str const buffer) {
+  return (az_json_state){
+    .buffer = buffer,
+    .i = 0,
+    .stack = 1,
+  };
+}
+
+az_result az_json_get_char(az_const_str const buffer, size_t const i, char *const out_char) {
+  if (i == buffer.size) {
+    return AZ_JSON_ERROR_UNEXPECTED_END;
+  }
+  *out_char = az_const_str_item(buffer, i);
+  return AZ_OK;
+}
+
+az_result az_json_expect_char(az_const_str const buffer, size_t *const p, char const expected) {
+  char c;
+  AZ_RETURN_ON_ERROR(az_json_get_char(buffer, *p, &c));
+  if (c != expected) {
+    return AZ_JSON_ERROR_UNEXPECTED_CHAR;
+  }
+  ++*p;
+  return AZ_OK;
+}
+
+az_result az_json_read_white_space(az_const_str const buffer, size_t *const p) {
+  for (; *p != buffer.size; ++*p) {
+    if (!az_json_is_white_space(az_const_str_item(buffer, *p))) {
+      break;
+    }
+  }
+  return AZ_OK;
+}
+
+az_result az_json_stack_push(az_json_state *const p_state, az_json_stack const stack) {
+  if (p_state->stack >> AZ_JSON_STACK_SIZE != 0) {
+    return AZ_JSON_ERROR_STACK_OVERFLOW;
+  }
+  p_state->stack = (p_state->stack << 1) | stack;
+  return AZ_OK;
+}
+
+az_result az_json_read_keyword_rest(
+  az_const_str const buffer,
+  size_t *const p,
+  az_const_str const keyword
+) {
+  ++*p;
+  for (size_t k = 0; k != keyword.size; ++*p, ++k) {
+    if (*p == buffer.size) {
+      return AZ_JSON_ERROR_UNEXPECTED_END;
+    }
+    if (az_const_str_item(buffer, *p) != az_const_str_item(keyword, k)) {
+      return AZ_JSON_ERROR_UNEXPECTED_CHAR;
+    }
+  }
+  return AZ_OK;
+}
+
+az_result az_json_read_number_digit_rest(
+  az_const_str const buffer,
+  size_t *const p,
+  double *const out_value,
```

Comment:
what does the standard actually say about how floating point values are read. I would like to see a TON of test cases for this function, preferably at least exhaustive testing for 32-bit floats

Reply:
As far as I know, JSON numer is not C number. Maybe, the JSON number set is a subset of C numbers.

Reply:
I mean the json standard. http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf

Reply:
Yes, the parser state-machine is implemented according to the standard. We need more tests, yes. We also need to make sure that it encodes the decimal floating-point number into binary floating-point number similar to C (or another standard).

Reply:
as far as I can tell there are few requirements on the number formats, although you do need to deal with all the possible ways to specify exponents and whatnot. charconv can definitely do it, and strtod / sscanf should be able to do a decent job as well.